### ISSUE TITLE
Unhelpful error message when using custom markdown processor

### ISSUE THREAD

**Issue by TWiStErRob** [jekyll/jekyll#3770]
Using 2.4.0 via github-pages gem v37.

`{% highlight java %}` is ugly (used to \`\`\` on GitHub issues) so I set `input:GFM` and then tried to integrate https://github.com/mvdbos/kramdown-with-pygments in `_plugins`.

This report is about the error experience I was subjected to.

### `jekyll serve --skip-initial-build`
```
Regenerating: 1 file(s) changed at 2015-06-09 19:52:40 Invalid Markdown Processor: KramdownPygments
                    Valid options are [ maruku | rdiscount | kramdown | redcarpet ]
  Conversion error: Jekyll::Converters::Markdown encountered an error converting '_posts/....md/#excerpt'.
  Conversion error: Invalid Markdown Processor: KramdownPygments
...error:
             Error: Invalid Markdown Processor: KramdownPygments
             Error: Run jekyll build --trace for more information.
```

### `jekyll serve` / `jekyll build --trace`
```
Invalid Markdown Processor: KramdownPygments
                    Valid options are [ maruku | rdiscount | kramdown | redcarpet ]
  Conversion error: Jekyll::Converters::Markdown encountered an error converting '_posts/....md/#excerpt'.
  Conversion error: Invalid Markdown Processor: KramdownPygments
             ERROR: YOUR SITE COULD NOT BE BUILT:
                    ------------------------------------
                    Invalid Markdown Processor: KramdownPygments
```

A few things to note here:
 1. It displayed a file I didn't change (for `--skip-initial-build`)  
That's kind of ok, since it's rebuilding dependents I guess
 1. The same error message repeated 3 times with many different formats:  
"Invalid Markdown Processor: KramdownPygments"
 1. Suggesting another command  
It's confusing to see a suggestion for `build` when I'm running `serve`. Granted, running a full build should give more info, but >>see next<<
 1. Suggesting something that's not possible: `--trace`  
I added the option and didn't make any difference, I also run the exact command suggested and it didn't give me more info other than "YOUR SITE COULD NOT BE BUILT".

The biggest issue is that it doesn't help at all to find out what the problem is.

After an hour of research I found that by copying `safe: true` from https://help.github.com/articles/using-jekyll-with-pages/#configuration-overrides I disabled parsing the .rb file I put in `_plugins` and that's why it doesn't work and not even suggested by [`third_party_processors`](https://github.com/jekyll/jekyll/blob/master/lib/jekyll/converters/markdown.rb)

**Comment by envygeeks** [comment]
I honestly don't know how we could phrase it better because as a programmer that message indicates to me immediately that my markdown processor is not available so a suggestion would be great as far as this is concerned because I guess it could be rather ambiguous.  It's made even more so clear by it spitting out the currently accepted Markdown processors... if it didn't then that is a bug.

The suggestion of trace is a default message just incase, but there was no error thrown so there could be no trace, we defensively program in Markdown checks rather than failing, you can only have a trace if there was an exception... there was no exception.

**Comment by parkr** [comment]
Can you re-do what you did and paste all input (not the site, just the commands) and output into a Gist and share with us that Gist? The descriptions are helpful, but it'd be nice to be more exact.

**Comment by TWiStErRob** [comment]
Guys, your response time and to-the-point-ness is great! Thanks! Sorry for nitpicking, but I hope you find some value in this report. Keep up the good work.

@parkr That's partially the point, there isn't much relevant output. But here's the [gist](https://gist.github.com/TWiStErRob/513c01b849e49f3285fb), I hope it matches what you wanted.

@envygeeks

> indicates to me immediately that my markdown processor is not available

It was really apparent that it wasn't available, what I couldn't grasp is why. I put the file in there and modified the config. "There's no more to it" - I though. I just started to build my site yesterday, so I'm *not* at home in this world yet (Ruby, Jekyll, Pygments, Perl)... and based on your home page ("Get up and running in seconds.") it should've been an easy thing, not a whole day.

> but there was no error thrown so there could be no trace

... but there's a `raise Errors::FatalException` there...

> spitting out the currently accepted Markdown processors

The problem is that it's only suggesting the internal ones, so from that it's not immediately implied it would suggest the custom ones, it could be a hard-coded error string. I discovered that it's dynamic after the 3rd skimming of the source. But most libraries/frameworks don't really help with listing good options, so that's a plus here!

#### Suggestion 1. "Run jekyll build --trace for more information."
This, for me, suggest that there'll be definitely going to (there's not even a "will" to soften it) be more information. Here's how Gradle does it:

> \* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output.

*Note: Gradle can't have a failure unless there was an Exception.* So I guess

> Try running `jekyll build --trace` to get the stack trace (if there's one).

would relax the suggestiveness a lot.

#### Suggestion 2. Make custom converters more user (~dev) friendly
We know that developers make errors. This is especially true for jekyll where the user might not be a university trained professional, only a technically inclined hobbyist.

There are multiple (non-exclusive) options here:
 * There's a huge warning at [Plugins](http://jekyllrb.com/docs/plugins/), but no mention at [Custom Markdown Processors](http://jekyllrb.com/docs/configuration/#custom-markdown-processors), not even a link to plugins which esentially a custom markdown processor is.
 * Also noticed a minor at [Configuration](http://jekyllrb.com/docs/configuration) > safe: "custom plugins, and ignore symbolic links." links to "Plugins" which doesn't contain anything about "symblic links".
 * Display `third_party_processors` separately. If the error message was

    > Invalid Markdown Processor: KramdownPygments, use one of the built in processors: [maruku, rdiscount, kramdown, redcarpet] or one of the custom processors: []

    It would've been more clear that the class is not being recognized.
 * Separate `allowed_custom_class` into multiple conditions and [guard](http://refactoring.com/catalog/replaceNestedConditionalWithGuardClauses.html) `self.class.const_get().new()` with those. Here's a really good (=bad) example of why that current `if` is insufficent:

    > Invalid Markdown Processor: My_Converter
Valid options are [ maruku | rdiscount | kramdown | redcarpet | My_Converter ]
 * In case you agree with the above: at `self.class.constants.include?` check or suggest to look at the `safe` configuration flag to see if that may be the reason why the class is not in the same module.

In general where I check for invalid data that stack frame is the one that knows and has the most infromation about the failure. By combining multiple checks into a blackbox method you lose a lot of information and one has to debug or interpret all the conditions one by one to figure out which went wrong. ... weirdly in my case both the classname and the package was correct, so I went for *WTF?* :)

**Comment by envygeeks** [comment]
> It was really apparent that it wasn't available, what I couldn't grasp is why. I put the file in there and modified the config. "There's no more to it" - I though. I just started to build my site yesterday, so I'm not at home in this world yet (Ruby, Jekyll, Pygments, Perl)... and based on your home page ("Get up and running in seconds.") it should've been an easy thing, not a whole day.

Well that should have been obvious because (and I'd have to go back and double check but it should) have mentioned that plugins are *not* loaded when safe mode is on, it should have been clearly documented on both Github's documentation as well as our own.  If it's not then please do let me know and I'll fix it here and submit a pull there.

> ... but there's a raise Errors::FatalException there...

You are right and I don't remember if RuntimeError is caught by a standard rescue.  This is more than likely a bug but this should be filed separately of this issue because the lack of more context and `--trace` while related in your case are unrelated in general.

> The problem is that it's only suggesting the internal ones, so from that it's not immediately implied it would suggest the custom ones, it could be a hard-coded error string. I discovered that it's dynamic after the 3rd skimming of the source. But most libraries/frameworks don't really help with listing good options, so that's a plus here!

Well this here presents itself with a logic problem.  We can't very well suggest something we don't know exists.  Now, if safe is off in theory it should show your custom processor and if it doesnt then again please file a separate issue because while related to your issue in general it's unrelated to your request to provide more context.

Yes, it is dynamic in a lot of ways and years ago when I worked with @parkr to build parts of that it was just slapped together (because I needed to use some custom processors I built for a client) and it certainly could use some reorganization (something I'm (almost) confident anybody on @jekyll/core would be willing to do.

---

What we could do is go back and clean up that source (something I'm willing to do if nobody else on @jekyll/core wants to) and provide a cleaner message.  First and foremost that double error needs to go, the raise need not duplicate a previously stated error, second thing we could do is insert a `logger.info` message hinting that safe might be a problem since we can readily detect safe mode, this would hint to you that hey your plugin isn't loading.  The third thing we could do is split up that mess that is setup.  Normally I don't advocate splitting up methods all over the place but we made a mess of that method and it could use some love.

**Comment by TWiStErRob** [comment]
> it should have been clearly documented on both Github's documentation as well as our own.

There's not a word "plugin" at [GitHub's doc](https://help.github.com/articles/using-jekyll-with-pages/#configuration-overrides) only a `safe: true` without explanation. I guess I should've went to find the documentation of those "Configuration Overrides" at http://jekyllrb.com/docs/configuration. So there's a missing warning in GitHub's docs or at least a link to jekyll docs (not to the home page though...).

As to Jekyll's docs see my two suggestions about that 1st and 2nd points in suggestion 2.

> if safe is off in theory it should show your custom processor

Don't worry there's no bug I know of in plugin discovery, except maybe if you include underscore in the class name or the class name is all lowercase. (4th point in suggestion 2) Should I open a separate issue for these?

> We can't very well suggest something we don't know exists

But you can suggest the lack of it (empty list), so it's clear you WOULD suggest if it was there. (3rd point in suggestion 2)

@envygeeks can you please file the spearate issues you may need as I probably can't provide enough context in the description (except for the strict class naming requirements issue).

**Comment by envygeeks** [comment]
On principle I cannot file issues on your behalf for several reasons:

* I am not you, I cannot provide your context.
* These are your suggestions not mine and filing them gives them my endorsement.
* If I do it for you I'll have to do it for everybody.

**Comment by TWiStErRob** [comment]
@envygeeks Added two new ones. Do you need one for those minor documentation changes too?

**Comment by parkr** [comment]
Is this fixed by #3771?

**Comment by TWiStErRob** [comment]
I like the logging changes, they reflect what @envygeeks said at the end of https://github.com/jekyll/jekyll/issues/3770#issuecomment-110528551, I think this could be closed, but try to give some love to the forked #3773 if you can.
